### PR TITLE
Buffer Package Size Reduction

### PR DESCRIPTION
**Pull request by noahdarveau-MSFT**
For more information about how to contribute to this repo, visit [this page](https://github.com/OfficeDev/microsoft-teams-library-js/blob/main/CONTRIBUTING.md).

## Description

> Summarize the changes, including the goals and reasons for this change. Be sure to call out any technical or behavior changes that reviewers should be aware of.

> If this Pull Request should close/resolve any issues when merged, use [the special syntax for that](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) here.

### Main changes in the PR:

1. We are currently using the [buffer npm package](https://www.npmjs.com/package/buffer) to polyfill the buffer node package when the clipboard api is being used in a web browser. The npm buffer package is quite sizable and we are not using all of its functionality, so I pulled out only the functionality we needed from the npm buffer package and made the skeleton-buffer package to replace using the full npm package. This isn't intended to be a permanent solution as if we make any further additions to the clipboard api, or need to use any other functionality from the buffer package in the future, we will likely either need to add that functionality to the skeleton-buffer package, or more likely, switch back to using the actual buffer npm package. However, in the meantime, this is a decent measure to reduce a good amount of package size until tree-shaking can be implemented. This should decrease the teams-js package size ~10% from the current main size

### A note on licensing
I just wanted to note that I did look into the licensing of the buffer npm package, and since it is under an MIT license and since teams-js is also under an MIT license, we should not run into any licensing problems with using and publishing a stripped version of their package.

### Validation performed:

1. <Step 1>
3. <Step 2>

### Unit Tests added:

> Unit tests are required for all changes. If no unit tests were added as part of this change, please explain why they aren't necessary.

<Yes/No>

### End-to-end tests added:

<Yes/No>

### Change file added:

> Ensure the change file meets the [formatting requirements](https://github.com/OfficeDev/microsoft-teams-library-js/blob/main/CONTRIBUTING.md#change-log-using-beachball).

<Yes/No>

### Related PRs:

> Remove this section if n/a

### Next/remaining steps:

> List the next or remaining steps in implementing the overall feature in subsequent PRs (or is the feature 100% complete after this?).

> Remove this section if n/a

- [ ] Item 1
- [ ] Item 2

### Screenshots:

> Remove this section if n/a

| Before     | After      |
| ---------- | ---------- |
| < image1 > | < image2 > |

### CODE REVIEW

File: packages/teams-js/package.json [hunk 1]
```diff
@@ -21,7 +21,8 @@
     "test:verbose": "pnpm test -- --verbose false --silent false"
   },
   "dependencies": {
-    "debug": "^4.3.3"
+    "debug": "^4.3.3",
+    "base64-js": "^1.3.1"
```
Comment (AE-MS) [inline packages/teams-js/package.json:25]: Can you alphabetize these?

File: change/@microsoft-teams-js-723b7387-886b-45e4-9191-9bc13fc83599.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+  "type": "patch",
+  "comment": "Made a skeletonized version of the buffer npm package to reduce bundle size",
```
Comment (AE-MS) [inline change/@microsoft-teams-js-723b7387-886b-45e4-9191-9bc13fc83599.json:3]: >  [](http://example.com/codeflow?start=67&length=1)

Perhaps worth inserted "as a temporary measure" here, based on your PR description?

**Comment by AE-MS** [comment]
>   /buffer@6.0.3:

Do you know what causes these references to buffer to still remain?

---
Refers to: pnpm-lock.yaml:5634 in deb060f. [](commit_id = deb060f4ef706198aba8f6ec62cdd60396faebbc, deletion_comment = False)

**Review by AE-MS — CHANGES_REQUESTED** [review]
:clock1:

**Comment by AE-MS** [comment]
Could you also create a work item tracking switching back to "real" buffer, or truly taking ownership of this, or something? I just don't want us to forget that we have a temporary copy of some code and aren't keeping up with security updates or something.

**Comment by noahdarveau-MSFT** [comment]
> > /buffer@6.0.3:
> 
> Do you know what causes these references to buffer to still remain?
> 
> Refers to: pnpm-lock.yaml:5634 in [deb060f](https://github.com/OfficeDev/microsoft-teams-library-js/commit/deb060f4ef706198aba8f6ec62cdd60396faebbc). [](commit_id = [deb060f](https://github.com/OfficeDev/microsoft-teams-library-js/commit/deb060f4ef706198aba8f6ec62cdd60396faebbc), deletion_comment = False)

Some of our dev dependencies use the buffer package as a dependency, so the references to the buffer package in the pnpm-lock file are not going to go away entirely. The difference is that the remaining buffer references are now under the devDependency section rather than the dependency section as it originally was. Since dev dependencies are not bundled into our package (unlike direct dependencies) it doesn't affect our bundle size anymore.

**Comment by noahdarveau-MSFT** [comment]
> Could you also create a work item tracking switching back to "real" buffer, or truly taking ownership of this, or something? I just don't want us to forget that we have a temporary copy of some code and aren't keeping up with security updates or something.

I went ahead and made a work item to return to the original buffer package once tree-shaking is implemented as well as to update the skeletonized buffer package if there are any updates to the buffer npm package in the meantime, and assigned it to myself.

**Review by TrevorJoelHarris — DISMISSED** [review]

**Review by AE-MS — APPROVED** [review]
:shipit:

**Review by AE-MS — APPROVED** [review]
:shipit:

**Review by TrevorJoelHarris — APPROVED** [review]

**Pull request merged**